src/promptflow-evals/promptflow/evals/evaluators/chat/__init__.py

```diff
@@ -0,0 +1,189 @@
+        """
+        Initialize an evaluator configured for a specific Azure OpenAI model.
+
+        :param model_config: Configuration for the Azure OpenAI model.
+        :type model_config: AzureOpenAIConnection
+        :param deployment_name: Deployment to be used which has Azure OpenAI model.
+        :type deployment_name: AzureOpenAIConnection
+        :return: A function that evaluates and generates metrics for "chat" scenario.
+        :rtype: function
+
+        **Usage**
+
+        .. code-block:: python
+
+            eval_fn = ChatEvaluator(model_config, deployment_name="gpt-4")
+            conversation = [
+                {"role": "user", "content": "What is the value of 2 + 2?"},
+                {"role": "assistant", "content": "2 + 2 = 4", "context": {
+                    "citations": [{"id": "math_doc.md", "content": "Information about additions: 1 + 2 = 3, 2 + 2 = 4"}]}}
+            ]
+            result = chat_eval(conversation=conversation)
+        """
+
+        # TODO: Need a built-in evaluator for retrieval. It needs to be added to `self._rag_evaluators` collection
+        self._rag_evaluators  = [
+            GroundednessEvaluator(model_config, deployment_name=deployment_name),
+            RelevanceEvaluator(model_config, deployment_name=deployment_name),
+        ]
+        self._non_rag_evaluators = [
+            CoherenceEvaluator(model_config, deployment_name=deployment_name),
+            FluencyEvaluator(model_config, deployment_name=deployment_name),
+        ]
+
+    def __call__(self, *, conversation: List[Dict], parallel=True, **kwargs):
+        """Evaluates chat scenario.
+
+        :param conversation: The conversation to be evaluated. Each turn should have "role" and "content" keys.
+                             "context" key is optional for assistant's turn and should have "citations" key with list of citations.
+        :type conversation: List[Dict]
+        :param parallel: If True, use parallel execution for evaluators. Else, use sequential execution. Default is True.
+        :type parallel: bool
+        :return: The scores for Chat scenario.
+        :rtype: dict
+        """
+
+        self._validate_conversation(conversation)
+
+        # Extract questions, answers and contexts from conversation
+        questions = []
+        answers = []
+        contexts = []
+        for turn_num, each_turn in enumerate(conversation):
+            role = each_turn["role"]
+            if role == "user":
+                questions.append(each_turn["content"])
+            elif role == "assistant":
+                answers.append(each_turn["content"])
+                if "context" in each_turn and "citations" in each_turn["context"]:
+                    citations = json.dumps(each_turn["context"]["citations"])
+                    contexts.append(citations)
+
+        # Select evaluators to be used for evaluation
+        compute_rag_based_metrics = True
+        if len(answers) != len(contexts):
```

Comment:
rag metrics and non-rag metrics require paired question-answer. May need to add the check if len(questions) == len(answers).

Reply:
If you navigate to `_validate_conversation` function, check out the code around variable `expected_role`,  this variable basically ensures the question and answer are coming in pairs.

Reply:
I read through the chat protocol, which allows roles e.g.  system, besides "user" and "assistant". Just wondering if this code need to handle this scenario (It may not be an issue for now.)

Reply:
Thanks for bringing it up. Current validation logic in `ChatEvaluator` just matches with the existing chat eval flow. If you look at the code below, "azureml-metrics" will fail the validation if the role is not in ["user", "assistant"] in any turn of the conversation. But we can always revisit this, and consider allowing other roles if there is an ask from customer. 

https://msdata.visualstudio.com/Vienna/_git/AzureMlCli?path=/src/azureml-metrics/azureml/metrics/common/_validation.py&version=GBmaster&line=939&lineEnd=940&lineStartColumn=1&lineEndColumn=108&lineStyle=plain&_a=contents